### PR TITLE
daphne: Implement fixed-size queries

### PR DESCRIPTION
**Pull request by cjpatton**
Partially addresses #100.
Based on #132 (merge that first).

Complete the implementation of fixed-size queries and add tests for MockAggregator. While at it, add the PartialBatchSelector struct to CollectResp.

### CODE REVIEW

File: daphne/src/lib.rs [hunk 1]
```diff
@@ -373,6 +399,29 @@ impl DapQueryConfig {
...
+/// the task's `time_precision` parameter; for fixed-size queries, the span consists of a single
+/// bucket, which is the batch determined by the batch ID (i.e., the partial batch selector).
+#[derive(Clone, Eq, Hash, PartialEq)]
+pub enum DapBatchBucket<'a> {
+    FixedSize { batch_id: &'a Id },
+    TimeInterval { batch_window: Time },
+}
+
+impl<'a> DapBatchBucket<'a> {
+    /// TODO(issue #100) Deprecate this once "fixed_size" support is added.
```
Comment (cjpatton) [inline daphne/src/lib.rs:415]: Reviewer note: This method is still needed by daphne_worker, which does not yet implement fixed-size queries.

File: daphne/src/lib.rs [hunk 2]
```diff
@@ -421,6 +470,87 @@ impl DapTaskConfig {
     pub(crate) fn truncate_time(&self, time: Time) -> Time {
         time - (time % self.time_precision)
     }
+
```
Comment (cjpatton) [inline daphne/src/lib.rs:473]: Reviewer note: This replaces https://github.com/cloudflare/daphne/pull/133/files#diff-e9132ff085252548c520081f56624153a00c9a0b0310982ae9a9aae165e6a9f3L606-R742

File: daphne/src/lib.rs [hunk 3]
```diff
@@ -496,6 +626,7 @@ pub struct DapLeaderUncommitted {
 /// The Helper's state during the aggregation flow.
 #[derive(Clone, Debug, PartialEq)]
 pub struct DapHelperState {
+    pub(crate) part_batch_sel: PartialBatchSelector,
```
Comment (cjpatton) [inline daphne/src/lib.rs:629]: Reviewer note: This structure carries state across aggregate requests. The Helper needs to be able to remember the batch ID indicated by the Leader in its AggregateInitializeReq.

File: daphne/src/roles_test.rs [hunk 4]
```diff
@@ -1286,3 +1322,24 @@ async fn e2e() {
         .query_for_current_batch_window(t.now);
     t.run_col_job(task_id, &query).await.unwrap();
 }
+
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:1325]: Reviewer note: There are a few more test cases we would like to add. They're on my TODO list -- right now I'm prioritizing getting to the point where we can start interop testing with others. That said, we could add them as part of this PR or document them in an issue.

File: daphne/src/testing.rs [hunk 5]
```diff
@@ -98,6 +112,76 @@ impl MockAggregator {
...
+                for (batch_id, report_count) in leader_state_store.batch_queue.iter_mut() {
+                    if *report_count < min_batch_size {
+                        *report_count += 1;
+                        return Some(DapBatchBucketOwned::FixedSize {
+                            batch_id: batch_id.clone(),
+                        });
+                    }
+                }
+
+                // No unsachurated batch exists, so crate a new batch.
```
Comment (armfazh) [inline daphne/src/testing.rs:148]:
```suggestion
                // No unsaturated batch exists, so crate a new batch.
```

**Review by armfazh — APPROVED** [review]

**Review by bhalleycf — APPROVED** [review]

**Pull request merged**